File: sdk/python/src/conduyt/modules/servo.py
"""CONDUYT Servo Module"""

from __future__ import annotations

import struct
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from conduyt.device import ConduytDevice

from conduyt.core.constants import CMD


class ConduytServo:
    def __init__(self, device: ConduytDevice, module_id: int = 0) -> None:
        self._device = device
        self._mod_id = module_id

    async def attach(self, pin: int, min_us: int = 544, max_us: int = 2400) -> None:
        payload = bytes([self._mod_id, 0x01, pin,
                         min_us & 0xFF, (min_us >> 8) & 0xFF,
                         max_us & 0xFF, (max_us >> 8) & 0xFF])
        await self._device._send_command(CMD.MOD_CMD, payload)

    async def write(self, angle: int) -> None:
        payload = bytes([self._mod_id, 0x02, angle & 0xFF])
        await self._device._send_command(CMD.MOD_CMD, payload)

    async def write_microseconds(self, us: int) -> None:
        payload = bytes([self._mod_id, 0x03, us & 0xFF, (us >> 8) & 0xFF])
        await self._device._send_command(CMD.MOD_CMD, payload)

    async def detach(self) -> None:
        payload = bytes([self._mod_id, 0x04])
        await self._device._send_command(CMD.MOD_CMD, payload)

File: sdk/swift/Sources/ConduytKit/Modules/Servo.swift
import Foundation

/// Typed wrapper for the CONDUYT Servo module.
@available(iOS 15.0, macOS 12.0, *)
public class ConduytServo {
    private let device: ConduytDevice
    private let moduleID: UInt8

    public init(device: ConduytDevice, moduleID: UInt8) {
        self.device = device
        self.moduleID = moduleID
    }

    /// Attach to a pin. minUs/maxUs of 0 fall back to firmware defaults (544/2400).
    public func attach(pin: UInt8, minUs: UInt16 = 0, maxUs: UInt16 = 0) async throws {
        var p = Data([moduleID, 0x01, pin])
        p.appendLE(uint16: minUs)
        p.appendLE(uint16: maxUs)
        _ = try await device.modCmd(p)
    }

    /// Write angle (0–180).
    public func write(angle: UInt8) async throws {
        _ = try await device.modCmd(Data([moduleID, 0x02, angle]))
    }

    /// Write a pulse width in microseconds.
    public func writeMicroseconds(_ us: UInt16) async throws {
        var p = Data([moduleID, 0x03])
        p.appendLE(uint16: us)
        _ = try await device.modCmd(p)
    }

    /// Detach the servo.
    public func detach() async throws {
        _ = try await device.modCmd(Data([moduleID, 0x04]))
    }
}
